### PR TITLE
alloc: avoid creating references in implementation of `Rc`/`Arc`

### PR DESCRIPTION
**Pull request by kadiwa4**
I noticed that the implementations of `Rc`/`Arc` create a lot of intermediate references to values that are not initialised or that shouldn't be accessed because of aliasing rules. This PR replaces some dubious cases with `ptr::addr_of(_mut)` and replaces `Layout::for_value` with `Layout::for_value_raw`.

### CODE REVIEW

File: library/alloc/src/rc.rs [hunk 1]
```diff
@@ -1987,15 +1991,15 @@ impl<T> Rc<[T]> {
...
             // Pointer to first element
-            let elems = &mut (*ptr).value as *mut [T] as *mut T;
+            let elems = ptr::addr_of_mut!((*ptr).value) as *mut T;
 
             let mut guard = Guard { mem: NonNull::new_unchecked(mem), elems, layout, n_elems: 0 };
 
-            for (i, item) in iter.enumerate() {
-                ptr::write(elems.add(i), item);
+            for item in iter {
+                ptr::write(elems.add(guard.n_elems), item);
```
Comment (kadiwa4) [inline library/alloc/src/rc.rs:2002]: drive-by simplification

**Comment by taiki-e** [comment]
I believe this is a duplicate of https://github.com/rust-lang/rust/pull/119433.

**Comment by kadiwa4** [comment]
whoops, I missed that while searching for existing issues :/
my PR has more changes however

File: library/alloc/src/rc.rs [hunk 2]
```diff
@@ -1628,7 +1628,7 @@ impl<T: ?Sized, A: Allocator> Rc<T, A> {
     pub unsafe fn get_mut_unchecked(this: &mut Self) -> &mut T {
         // We are careful to *not* create a reference covering the "count" fields, as
         // this would conflict with accesses to the reference counts (e.g. by `Weak`).
-        unsafe { &mut (*this.ptr.as_ptr()).value }
+        unsafe { &mut *ptr::addr_of_mut!((*this.ptr.as_ptr()).value) }
```
Comment (Nilstrieb) [inline library/alloc/src/rc.rs:1631]: I don't think this created any unnecessary references, the `addr_of_mut` roundtrip seems pointless.

Comment (kadiwa4) [reply library/alloc/src/rc.rs:1631]: Maybe you're right, but this does result in different MIR output.
I thought that they were different because the previous line created a reference to the entire struct, not just that one field

Comment (taiki-e) [reply library/alloc/src/rc.rs:1631]: > I thought that they were different because the previous line created a reference to the entire struct, not just that one field

IIUC, `&mut (*ptr).field` does not create a reference to the entire struct. See also the following lines in Weak::inner.

https://github.com/rust-lang/rust/blob/87e1430893d55001034bd76c5dbe832d80bc50c3/library/alloc/src/sync.rs#L2849-L2852

Comment (kadiwa4) [reply library/alloc/src/rc.rs:1631]: I saw that comment but I didn't quite trust it because it's apparently part of faulty code :) I guess I can close this tho
